### PR TITLE
repository_ctx: Allow renaming archive entries during extraction.

### PR DESCRIPTION
**Pull request by jmillikin**
Adds a `rename_files=` parameter to the `extract()` and `download_and_extract()` methods of `repository_ctx`. This new parameter takes a dict of archive entries to rename, and is applied prior to any prefix adjustment.

### CODE REVIEW

**Comment by jmillikin** [comment]
Use cases include:
* Archives where multiple files have the same name in different case. On platforms with case-insensitive filesystems (macOS, Windows) these archives can't be extracted as-is, and the files can't be fixed up after extraction.
  * One such project is the Linux kernel, which has files named like `net/netfilter/xt_dscp.c` and `net/netfilter/xt_DSCP.c` (among others).
* Archives containing non-UTF8 filenames fail to extract on platforms or filesystems that enforce Unicode paths.
  * See https://github.com/bazelbuild/rules_go/issues/2771 (Go toolchains from golang.org fail to extract)

A longer example of extracting the Linux kernel on macOS (which is case-insensitive by default):
```python
load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")

def _linux_repository(ctx):
    ctx.file("WORKSPACE", "workspace(name = {name})\n".format(
        name = repr(ctx.name),
    ))
    ctx.file("BUILD.bazel", """
genrule(
    name = "nf_srcs",
    srcs = glob(["net/netfilter/*"]),
    outs = ["nf_srcs.txt"],
    cmd = "ls -1 external/linux/net/netfilter/ > $@",
)
""")

    ctx.download_and_extract(
        url = ["https://cdn.kernel.org/pub/linux/kernel/v5.x/linux-5.19.tar.xz"],
        sha256 = "ff240c579b9ee1affc318917de07394fc1c3bb49dac25ec1287370c2e15005a8",
        stripPrefix = "linux-5.19",
        renameFiles = _rename_files("5.19"),
    )

def _rename_files(version):
    netfilter_srcs = {
        "xt_DSCP.c": "xt_DSCP_target.c",
        "xt_HL.c": "xt_HL_target.c",
        "xt_RATEEST.c": "xt_RATEEST_target.c",
        "xt_TCPMSS.c": "xt_TCPMSS_target.c",
    }
    renames = {}
    for orig_name, new_name in netfilter_srcs.items():
        prefix = "linux-{}/net/netfilter/".format(version)
        renames[prefix + orig_name] = prefix + new_name
    # and similar logic for netfilter headers under include/uapi/linux/
    return renames

linux_repository = repository_rule(
    implementation = _linux_repository,
)
```

```
$ bazel build @linux//:nf_srcs
[...]
Target @linux//:nf_srcs up-to-date:
  bazel-bin/external/linux/nf_srcs.txt
$ cat bazel-bin/external/linux/nf_srcs.txt | grep -i dscp
xt_DSCP_target.c
xt_dscp.c
```

File: src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkRepositoryContext.java [hunk 1]
```diff
@@ -749,8 +749,22 @@ public StructImpl download(
                     + " archive. Instead of needing to specify this prefix over and over in the"
                     + " <code>build_file</code>, this field can be used to strip it from extracted"
                     + " files."),
+        @Param(
+            name = "renameFiles",
```
Comment (fmeum) [inline src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkRepositoryContext.java:753]: This is the first time I actually noticed that `stripPrefix` has a camel case name - that looks like a bug as Starlark parameters usually use snake case. Using snake case for `renameFiles` at least wouldn't complicate future efforts to clean up this inconsistency.

Comment (jmillikin) [reply src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkRepositoryContext.java]: Done.

**Comment by jmillikin** [comment]
Gentle ping

File: src/main/java/com/google/devtools/build/lib/bazel/repository/DecompressorDescriptor.java [hunk 2]
```diff
@@ -33,17 +34,21 @@ public class DecompressorDescriptor {
...
   private final Optional<String> prefix;
   private final boolean executable;
+  private final Map<String, String> renameFiles;
   private final Decompressor decompressor;
 
   private DecompressorDescriptor(
       String targetKind, String targetName, Path archivePath, Path repositoryPath,
-      @Nullable String prefix, boolean executable, Decompressor decompressor) {
+      @Nullable String prefix, boolean executable,
+      @Nullable Map<String, String> renameFiles,
```
Comment (Wyverald) [inline src/main/java/com/google/devtools/build/lib/bazel/repository/DecompressorDescriptor.java:43]: this doesn't need to be nullable (it's never null anyway, and null has the same semantics as empty). also remove all the null checks

File: src/main/java/com/google/devtools/build/lib/bazel/repository/DecompressorDescriptor.java [hunk 3]
```diff
@@ -164,6 +175,12 @@ public Builder setExecutable(boolean executable) {
       return this;
     }
 
+    @CanIgnoreReturnValue
+    public Builder setRenameFiles(Map<String, String> renameFiles) {
+      this.renameFiles = renameFiles;
```
Comment (Wyverald) [inline src/main/java/com/google/devtools/build/lib/bazel/repository/DecompressorDescriptor.java:180]: defensively copy here (ImmutableMap.copyOf)

File: src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkRepositoryContext.java [hunk 4]
```diff
@@ -749,8 +749,22 @@ public StructImpl download(
...
                     + " files."),
+        @Param(
+            name = "rename_files",
+            defaultValue = "{}",
+            named = true,
+            positional = false,
+            doc =
+                "An optional dict specifying files to rename during the extraction. Archive entries"
+                    + " with names exactly matching a key will be renamed to the value, prior to"
+                    + " any directory prefix adjustment."),
```
Comment (Wyverald) [inline src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkRepositoryContext.java:760]: could you add more information on how this parameter can be used, like you did in the PR comments? or maybe even link to this PR

File: src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkRepositoryContext.java [hunk 5]
```diff
@@ -881,6 +900,15 @@ public void extract(Object archive, Object output, String stripPrefix, StarlarkT
...
                     + " easier but should be set before shipping."),
+        @Param(
+            name = "rename_files",
+            defaultValue = "{}",
+            named = true,
+            positional = false,
+            doc =
+                "An optional dict specifying files to rename during the extraction. Archive entries"
+                    + " with names exactly matching a key will be renamed to the value, prior to"
+                    + " any directory prefix adjustment."),
```
Comment (Wyverald) [inline src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkRepositoryContext.java:911]: same here

File: src/test/java/com/google/devtools/build/lib/bazel/repository/CompressedTarFunctionTest.java [hunk 6]
```diff
@@ -64,6 +66,26 @@ public void testDecompressWithPrefix() throws Exception {
     archiveDescriptor.assertOutputFiles(outputDir, INNER_FOLDER_NAME);
   }
 
+  /**
+   * Test decompressing a tar.gz file, with some entries being renamed during the
+   * extraction process.
+   */
+  @Test
+  public void testDecompressWithRenamedFiles() throws Exception {
```
Comment (Wyverald) [inline src/test/java/com/google/devtools/build/lib/bazel/repository/CompressedTarFunctionTest.java:74]: can we add a test case where strip_prefix is also present, to test that rename_files happens before strip_prefix?

**Review by Wyverald — CHANGES_REQUESTED** [review]
thanks!

File: src/main/java/com/google/devtools/build/lib/bazel/repository/DecompressorDescriptor.java (hunk 3 above)
Comment (jmillikin) [reply src/main/java/com/google/devtools/build/lib/bazel/repository/DecompressorDescriptor.java]: Done.

File: src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkRepositoryContext.java (hunk 5 above)
Comment (jmillikin) [reply src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkRepositoryContext.java]: Done.

File: src/test/java/com/google/devtools/build/lib/bazel/repository/CompressedTarFunctionTest.java (hunk 6 above)
Comment (jmillikin) [reply src/test/java/com/google/devtools/build/lib/bazel/repository/CompressedTarFunctionTest.java:74]: Done.

**Review by Wyverald — APPROVED** [review]
thanks!

**Pull request closed without merging**